### ISSUE TITLE
Kodi GUI freeze at boot on SFTP timeout, and unwanted library behaviour

### ISSUE THREAD

**Issue by jobhh** [xbmc/xbmc#15727]
I've been thinking about posting a bug report/feature request for a couple years. Finally the time has come to do so :)

### Describe the bug
An initial SFTP connection timeout when Kodi is started results in unwanted behaviour in GUI and library.

## Expected Behaviour
Kodi should start, connect to all sources while remaining responsive, inform the user of the progress and/or issues and be ready for use.

## Actual Behaviour
When starting Kodi, while the Synology NAS is in low power mode, Kodi has a SFTP connection timeout before the Synology is booted up and ready to respond. Until this timeout the Kodi GUI remains completely frozen, for about 10 seconds. When after some time a movie is selected the SFTP session remains "Not connected", even though the Synology is booted up at this point and ready to stream. The progress of a selected movie is lost, probably because Kodi assumes it doesn't exist anymore.

## Possible Fix
If the GUI and SFTP connection share the same thread this might be difficult... But this is my wish list:

1. A notice like "connecting to source: SFTP:\\etc etc" in the top right is more informative than a 10 second system hang.
2. When a timeout occurs, or a source isn't available in general (usb/smb/sftp/etc), this should be communicated through the GUI.
3. It should be possible to attempt another connection to a source without restarting Kodi. Right now the SFTP connection remains not connected, even if a movie from the SFTP source is selected and the SFTP source itself is ready for use by that time.
4. If a source is still not available when a movie is selected, ask for user input. Don't automatically assume the movie can be deleted from the "in progress movies" list.
5. In addition to 4, it would be nice to have movies from a currently unavailable source greyed out on the homescreen. When a connection can be made the movie images can be normal again, which also seems like a clear (addition) cue a connection has been established.

### To Reproduce
Steps to reproduce the behaviour:

1. Enable SFTP plugin
2. Add SFTP source
3. Index SFTP source
4. Exit Kodi
4. Wait for the SFTP source to sleep (Synology NAS in my case)
5. Start Kodi. At this point the SFTP source will start to wake up, since Kodi is trying to connect to a port, but too slowly.
6. Watch GUI freeze.
7. Select an indexed movie and watch its progress disappear, or just not start if no progress existed.

### Debuglog
The debuglog can be found here, this are the (redacted) relevant lines:

    21:02:38.361 T:19212   ERROR: AddOnLog: SFTP support: SFTPSession: Failed to connect 'Timeout connecting to HOSTNAME'
    21:02:47.984 T:12008  NOTICE: VideoPlayer::OpenFile: sftp://HOSTNAME:PORT/Media/Movies/DEADPOOLFOLDER/DEADPOOLMOVIE
    21:02:47.987 T:4968  NOTICE: Creating InputStream
    21:02:47.987 T:4968   ERROR: AddOnLog: SFTP support: SFTPSession: Not connected, can't list directory 'Media/Movies/DEADPOOLFOLDER/'
    21:02:47.987 T:4968   ERROR: XFILE::CDirectory::GetDirectory - Error getting sftp://HOSTNAME:PORT/Media/Movies/DEADPOOLFOLDER/
    21:02:47.988 T:4968   ERROR: AddOnLog: SFTP support: SFTPSession: Not connected, can't list directory 'Media/Movies/DEADPOOLFOLDER/'
    21:02:47.988 T:4968   ERROR: XFILE::CDirectory::GetDirectory - Error getting sftp://HOSTNAME:PORT/Media/Movies/DEADPOOLFOLDER/
    21:02:47.989 T:4968   ERROR: AddOnLog: SFTP support: SFTPSession: Not connected and can't create file handle for 'Media/Movies/DEADPOOLFOLDER/DEADPOOLMOVIE'
    21:02:47.989 T:4968   ERROR: XFILE::CFileCache::Open - failed to open source <sftp://HOSTNAME:PORT/Media/Movies/DEADPOOLFOLDER/DEADPOOLMOVIE>
    21:02:47.989 T:4968   ERROR: CVideoPlayer::OpenInputStream - error opening [sftp://HOSTNAME:PORT/Media/Movies/DEADPOOLFOLDER/DEADPOOLMOVIE]
    21:02:47.989 T:4968  NOTICE: CVideoPlayer::OnExit()
    21:02:48.112 T:12008  NOTICE: CVideoPlayer::CloseFile()
    21:02:48.113 T:12008  NOTICE: VideoPlayer: waiting for threads to exit
    21:02:48.113 T:12008  NOTICE: VideoPlayer: finished waiting

### Screenshots 
Here is a screen capture to help explain the problem:
https://gfycat.com/UnequaledHauntingAsp

### Your Environment
Used Operating system:
 - [x] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 10 Home 64bit version 1809
 - Kodi version: Kodi 18.1 Leia 64bit (not windows store, screen capture is from 18.0)
 - SFTP plugin version: 1.0.1

Issue is also present on Nvidia shield and for the past years. So I don't think it's OS related ;) 

Edit:
I've mentioned this behaviour in the SFTP VFS addon repo as well last year:
https://github.com/xbmc/vfs.sftp/issues/18#issuecomment-434456912
https://github.com/xbmc/vfs.sftp/pull/19#issuecomment-433648742

Edit 2:
The workaround I've been using the last years is to use my smartphone to connect to the NAS 30 seconds before starting Kodi on my media player. This way I know the NAS is properly awake before starting Kodi.

**Comment by quicksilvered** [comment]
I also experience similar behavior, albeit with a QNAP NAS, which is presumably on standby.

Additionally, and I believe it is related, if a NAS connection is lost (due to pulled network cable or NAS restart) - instead of expected behavior described above, Kodi UI completely freezes and does not accept attempts to be force closed (I was watching something on Kodi from NAS, updated NAS firmware, restarted NAS and forgot about Kodi stream - Kodi proceeded to lock-up and would not respond nor could it be force-closed)
